### PR TITLE
[enhance](io) Set segment file size to rowset meta

### PR DESCRIPTION
**Pull request by platoneko**
## Proposed changes

Set segment file size to rowset meta, to avoid the IO of fetching file meta info

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

File: be/src/olap/rowset/vertical_beta_rowset_writer.h [hunk 1]
```diff
@@ -17,28 +17,28 @@
...
 namespace vectorized {
 class Block;
 } // namespace vectorized
 
 // for vertical compaction
-// TODO(plat1ko): Inherited from template type `T`, `T` is `BetaRowsetWriter` or `CloudBetaRowsetWriter`
-class VerticalBetaRowsetWriter final : public BetaRowsetWriter {
+template <class T>
+    requires std::is_base_of_v<BaseBetaRowsetWriter, T>
+class VerticalBetaRowsetWriter final : public T {
```
Comment (gavinchou) [inline be/src/olap/rowset/vertical_beta_rowset_writer.h:36]: does gcc compile?

File: be/src/olap/rowset/rowset_meta.h [hunk 2]
```diff
@@ -314,6 +314,12 @@ class RowsetMeta {
 
     void set_txn_expiration(int64_t expiration) { _rowset_meta_pb.set_txn_expiration(expiration); }
 
+    // `seg_file_size` MUST ordered by segment id
+    void set_segments_file_size(const std::vector<size_t>& seg_file_size);
```
Comment (gavinchou) [inline be/src/olap/rowset/rowset_meta.h:318]: how to ensure 
> `seg_file_size` MUST ordered by segment id

should we check the order or sort it in `set_segments_file_size`

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 3]
```diff
@@ -96,12 +97,92 @@ void build_rowset_meta_with_spec_field(RowsetMeta& rowset_meta,
...
+        DCHECK(false);
+        return ResultError(Status::InternalError("get segments file size without closed"));
+    }
+
+    Status st;
+    std::vector<size_t> seg_file_size(_file_writers.size(), 0);
+    bool succ = std::all_of(_file_writers.begin(), _file_writers.end(), [&](auto&& it) {
+        auto&& [seg_id, writer] = it;
+
+        if (seg_id > seg_file_size.size()) [[unlikely]] {
```
Comment (gavinchou) [inline be/src/olap/rowset/beta_rowset_writer.cpp:142]: it seems to be `seg_id >= seg_file_size.size()` ?

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -96,12 +97,92 @@ void build_rowset_meta_with_spec_field(RowsetMeta& rowset_meta,
...
+                    fmt::format("invalid segment fsize={} path={}", fsize, writer->path().native());
+            DCHECK(false) << err_msg;
+            st = Status::InternalError(err_msg);
+            return false;
+        }
+
+        return true;
+    });
+
+    if (succ) {
```
Comment (gavinchou) [inline be/src/olap/rowset/beta_rowset_writer.cpp:172]: try ternary expr. to reduce lines of code

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 5]
```diff
@@ -664,6 +739,13 @@ Status BaseBetaRowsetWriter::_build_rowset_meta(RowsetMeta* rowset_meta, bool ch
     // TODO write zonemap to meta
     rowset_meta->set_empty((num_rows_written + _num_rows_written) == 0);
     rowset_meta->set_creation_time(time(nullptr));
+
+    if (auto seg_file_size = _seg_files.segments_file_size(); !seg_file_size.has_value()) {
+        LOG(ERROR) << seg_file_size.error();
```
Comment (gavinchou) [inline be/src/olap/rowset/beta_rowset_writer.cpp:744]: Should we return an error status?

**Review by kaijchen — APPROVED** [review]
LGTM

**Review by gavinchou — APPROVED** [review]

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**